Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp

```diff
@@ -115,8 +117,54 @@ void MappingButton::Detect()
     // Avoid that the button press itself is registered as an event
     Common::SleepCurrentThread(100);
 
-    const auto expr = MappingCommon::DetectExpression(
-        m_reference, dev.get(), m_parent->GetController()->GetDefaultDevice());
+    std::vector<std::future<QPair<QString, QString>>> futures;
+
+    QString expr;
+
+    if (m_parent->GetParent()->IsMappingAllDevices())
```

Comment:
This block from here down could conceivably go within its own function that returns the built-up string, which would let you do:

```cpp
const QString expr = MakeDeviceExpression(...);
```